Refs #20550 - Reset remote database

mbacovsky opened the pull request:
We usually don't have permissions to drop remote DBs. This fix lets installer to drop only tables and keep the empty DB during `--reset`. The original behavior (drop whole DB) is preserved for local databases.

https://github.com/Katello/katello-packaging/pull/499 is required for it to work

mbacovsky commented:
@ekohl updated. Is this acceptable solution for possible missing utils script?

ekohl commented on hooks/pre/10-reset_feature.rb:
```diff
@@ -25,7 +33,11 @@ def stop_services
   Kafo::KafoConfigure.logger.info 'Dropping database!'
-  Kafo::Helpers.execute('foreman-rake db:drop 2>&1')
+  if @db_config && @db_config.remote_db?(@db_config.foreman)
+    @db_config.empty_database!(@db_config.foreman)
```
This is me thinking out loud. There is ```db:purge``` since Rails 4.2 so I wonder if we should try to use that instead. On the other hand, we already need it because candlepin has no equivalent. Let's keep this version.

mbacovsky replied:
@ekohl thanks for pointing the `db:purge` out. It didn't came to my mind. I'm okay to use it for the foreman db if that helps. No strong opinion though.

mbacovsky commented:
@ekohl I see your point with the inclusion. Once the backup and restore are moved to foreman-maintain we can move the included script in this repo. I guess to copy it in here now would not win us much.

mbacovsky commented:
@ekohl, thanks for the review!

mbacovsky commented:
@ekohl is there anything I can help with to get this merged?

ekohl commented:
@mbacovsky I reviewed the packaging PR but I'm still not very happy with the design and dependencies of the ruby files.

mbacovsky commented:
I checked the `db:purge` and as it internally use drop [1] we can't use this to reset the remote db. I'll update the PR and handle remote DB manually. 

[1] https://github.com/rails/rails/blob/master/activerecord/lib/active_record/tasks/postgresql_database_tasks.rb#L45-L49

mbacovsky commented:
@ekohl, I've updated this PR so that it doesn't depend on the Katello packaging. Is it acceptable now?

mbacovsky commented:
I've added support for remote Mongo in here for easier testing. It is in separate commit.

johnpmitsch commented:
@ekohl has asked @mbacovsky to find another reviewer since he will be away (from #theforeman-dev)
```
[4/18/18 16:28] <ewoud> mbacovsk: I noticed you asked for my review on remote db in katello-installer ;
I'll check after my PTO but feel free to find another reviewer
```
Since myself and another reviewer found no issues with this PR, I'm going to merge so this change makes the branching date tomorrow.

@mbacovsky can you squash the commits?

mbacovsky commented on hooks/pre/10-reset_feature.rb:
```diff
@@ -35,26 +70,95 @@ def reset_candlepin
+  Kafo::Helpers.execute(cmd)
 end
 
 def reset_pulp
```
Good point, I'll file an issue for this.

johnpmitsch commented:
@ekohl You aren't good at PTO :wink: Once those issues are addressed would consider this okay to merge? (After I retest functionality)

mbacovsky commented on hooks/pre/10-reset_feature.rb:
```diff
@@ -35,26 +70,95 @@ def reset_candlepin
+  system(pg_sql_statement(config, delete_statements)) if delete_statements
+end
+
+def param_value(mod, name)
```
@ekohl, Any idea where to put it? I tried helpers hook, but I had to pass kafo instance to get to the params. Ideally this method should go Kafo::HookContext. Would it be okay to fix it as an separate issue?

mbacovsky commented:
I've fixed the seeds. I didn't found better solution for the `param_value` and would suggest to add this to Kafo::HookContext in separate PR. I can issue for this plus issue for not stopping pulp workers

ekohl commented:
My flight was delayed and I was bored. Once addressed 👍

> On 18 Apr 2018, at 18:33, John Mitsch <notifications@github.com> wrote:
> 
> @ekohl You aren't good at PTO 😉 Once those issues are addressed would consider this okay to merge? (After I retest functionality)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ekohl commented:
Separate issue is fine. I think we should extend kafo itself instead of doing it in hooks

> On 18 Apr 2018, at 18:48, mbacovsky <notifications@github.com> wrote:
> 
> @mbacovsky commented on this pull request.
> 
> In hooks/pre/10-reset_feature.rb:
> 
> > +def pg_sql_statement(config, statement)
> +  pg_command_base(config, 'psql', "-d #{config[:database]} -t -c \"" + statement + '"')
> +end
> +
> +# WARNING: deletes all the data from a database. No warnings. No confirmations.
> +def empty_database!(config)
> +  generate_delete_statements = pg_sql_statement(config, %q(
> +        select string_agg('drop table if exists \"' || tablename || '\" cascade;', '')
> +        from pg_tables
> +        where schemaname = 'public';
> +      ))
> +  delete_statements = `#{generate_delete_statements}`
> +  system(pg_sql_statement(config, delete_statements)) if delete_statements
> +end
> +
> +def param_value(mod, name)
> @ekohl, Any idea where to put it? I tried helpers hook, but I had to pass kafo instance to get to the params. Ideally this method should go Kafo::HookContext. Would it be okay to fix it as an separate issue?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

mbacovsky commented on hooks/pre/10-reset_feature.rb:
```diff
@@ -35,26 +70,95 @@ def reset_candlepin
+
+def load_mongo_config
+  config = {}
+  seeds = param_value('katello', 'pulp_db_seeds')
```
Fixed

alda519 replied:
this is probably wrong, when you have mongo replica set, you should to connect to primary node, because secondary will not allow you to write
```
# mongo --host 192.168.122.93:27017
replica1:SECONDARY> db.dropDatabase()
{ "note" : "from execCommand", "ok" : 0, "errmsg" : "not master" }
```
it cannot be guaranteed that the first one is primary, and the first can be down

you need url like `replica_name/host1:port,host2:port,...` to connect to primary
```
# mongo --host replica1/192.168.122.93:27017,192.168.122.35:27017
replica1:PRIMARY> use rere
switched to db rere
replica1:PRIMARY> db.dropDatabase()
{ "dropped" : "rere", "ok" : 1 }
```

mbacovsky replied:
@alda519, good point.
 
The question is - do we support replicas, do we want to block this PR on it?

mbacovsky replied:
I tried to look into this and it will need more investigation. According to some SO [1] comments the `--host` params can differ on 3.2 and 3.4+. I need to get more familiar with the setup and we should decide on which Mongo DB versions we want to support and test. I'll file a bug to track this.

mbacovsky replied:
http://projects.theforeman.org/issues/23337

mbacovsky commented on hooks/pre/10-reset_feature.rb:
```diff
@@ -23,9 +23,44 @@ def stop_services
 
+def load_foreman_config
+  db_config = {}
+  db_config[:host] = param('foreman', 'db_host').value || 'localhost'
```
Updated

mbacovsky commented:
I filed an issue for adding `param_value` to Kafo [1] and updated the helpers to monkey patch the context in the meantime.
I've added the step stopping the workers.

Now I consider all the issues addressed.

[1] http://projects.theforeman.org/issues/23332

alda519 commented:
mongo need to also use `--sslPEMKeyFile` if server wants to verify client's cert

problem is that is can be split in two files
```
/etc/pulp/server.conf
ssl_keyfile: /etc/ssl/mongo-client.key
ssl_certfile: /etc/ssl/mongo-client.crt
```

I will test postgres SSL too

mbacovsky commented:
@alda519 I'll update the PR to use ----sslPEMKeyFile if set according to the  [1]

[1] https://docs.mongodb.com/manual/tutorial/configure-ssl-clients/#connect-to-mongodb-instance-that-validates-when-presented-with-a-certificate

mbacovsky commented:
I've updated to support the client cert according to the docs

johnpmitsch commented:
@alda519 are you ok with the recent changes?

alda519 commented:
Yes I am. Drop mongo over SSL now works:
```
MongoDB shell version: 2.6.12
connecting to: 192.168.122.93:27017/pulp1db
[object Object]
mongo -u pulp1 -p pulp1pw --host 192.168.122.93 --port 27017 --ssl --sslCAFile /etc/ssl/ca.pem --sslPEMKeyFile /etc/ssl/mongo-client.pem --eval "db.dropDatabase();" pulp1db finished successfully!
````
And there is no problem with postgres and SSL either.

mbacovsky commented:
Thanks for testing @alda519!

johnpmitsch commented:
still works well and all concerns appear to be addressed so merging!
